src/metabase/api/dashboard.clj

```diff
@@ -643,4 +648,126 @@
+  {:arglists '([& {:keys [export-format parameters ignore_cache constraints parameters middleware]}])}
+  [& {:keys [dashboard-id card-id parameters export-format]
+      :or   {export-format :api}
+      :as   options}]
+  ;; make sure we can read this Dashboard. Card will get read-checked later on inside
+  ;; [[api.card/run-query-for-card-async]]
+  (api/read-check Dashboard dashboard-id)
+  (check-card-is-in-dashboard card-id dashboard-id)
+  (let [params  (resolve-params-for-query dashboard-id card-id parameters)
+        options (merge
+                 {:ignore_cache false
+                  :constraints  constraints/default-query-constraints
+                  :context      :dashboard}
+                 options
+                 {:parameters   params
+                  :dashboard-id dashboard-id})]
+    (m/mapply api.card/run-query-for-card-async card-id export-format options)))
+
+(def ParameterWithID
+  "Schema for a parameter map with an string `:id`."
+  (su/with-api-error-message
+    {:id       su/NonBlankString
+     s/Keyword s/Any}
+    "value must be a parameter map with an 'id' key"))
+
+(api/defendpoint POST "/:dashboard-id/card/:card-id/query"
+  "Run the query associated with a Saved Question (`Card`) in the context of a `Dashboard` that includes it."
+  [dashboard-id card-id :as {{:keys [parameters], :as body} :body}]
+  {parameters (s/maybe [ParameterWithID])}
+  (m/mapply run-query-for-dashcard-async
+            (merge
+             body
+             {:dashboard-id dashboard-id
+              :card-id      card-id})))
+
+(api/defendpoint POST "/:dashboard-id/card/:card-id/query/:export-format"
+  "Run the query associated with a Saved Question (`Card`) in the context of a `Dashboard` that includes it, and return
+  its results as a file in the specified format.
+
+  `parameters` should be passed as query parameter encoded as a serialized JSON string (this is because this endpoint
+  is normally used to power 'Download Results' buttons that use HTML `form` actions)."
+  [dashboard-id card-id export-format :as {{:keys [parameters], :as request-parameters} :params}]
+  {parameters    (s/maybe su/JSONString)
+   export-format api.dataset/ExportFormat}
+  (m/mapply run-query-for-dashcard-async
+            (merge
+             request-parameters
+             {:dashboard-id  dashboard-id
+              :card-id       card-id
+              :export-format export-format
+              :parameters    (json/parse-string parameters keyword)
+              :constraints   nil
+              ;; TODO -- passing this `:middleware` map is a little repetitive, need to think of a way to not have to specify
+              ;; this all over the codebase any time we want to do a query with an export format. Maybe this should be the
+              ;; default if `export-format` isn't `:api`?
+              :middleware    {:process-viz-settings?  true
```

Comment:
what if we had middleware groups? you can specify middleware and middleware groups and those would get expanded correctly?

Reply:
that could be a good idea. Might make sense as a follow-on refactor